apollo-router-core/src/layers/cache.rs

```diff
@@ -41,80 +46,92 @@ where
-                            cache.insert(key, Ok(value));
-                        }
-                        Err(err) => {
-                            cache.insert(key, Err(err.to_string()));
-                        }
+                // The key function returned none, so the request was uncachable.
+                self.inner.call(request).err_into().boxed()
+            }
+            Some(key) => {
+                let value = self.cache.get(key);
+                match value {
+                    None => {
+                        // This is a new value. Create a sentinel and lock it.
+                        // The sentinel is an RWLock over an initially empty option.
+                        // Once the sentinel is created it is locked it so that readers must wait to read.
+                        // The sentinel is then put into the cache so that any subsequent read
+                        // will pick this up and have to wait.
+                        let sentinel = Arc::new(RwLock::new(None));
+                        let mut guard = futures::executor::block_on(sentinel.clone().write_owned());
+                        self.cache.insert(key.clone(), sentinel);
+
+                        let value_fn = self.value_fn;
+                        self.inner
+                            .call(request)
+                            .err_into()
+                            .then(move |response| async move {
+                                // We got the response now. Update the sentinel and release the guard.
+                                let sentinel_value = match &response {
+                                    Ok(response) => Ok((value_fn)(response).clone()),
+                                    Err(e) => Err(e.to_string()),
+                                };
+                                let _ = guard.insert(sentinel_value);
+                                response
+                            })
+                            .boxed()
                     }
-                    response
-                })
+                    Some(value) => {
+                        // The value has been populated, we need to await on the sentinel.
+                        // The read lock on the sentinel is released when it is populated allowing the
+                        // current read to proceed.
+                        let response_fn = self.response_fn;
+                        value
+                            .read_owned()
+                            .map(move |value| {
+                                let value = value.clone().expect("Value will always have been set");
+                                match value {
+                                    Ok(value) => Ok((response_fn)(request, value)),
+                                    Err(err) => Err(err.into()),
+                                }
+                            })
+                            .boxed()
+                    }
+                }
```

Comment:
It's not that I dislike the sentinel approach, but it's a different way to do exactly the same thing that the existing logic we had in the de-duplication layer or the CachingMap. I think we should only do this kind of thing one way unless there's a compelling reason why we need to diverge.

Reply:
The primary reason is to remove the need for the service to clone as it doesn't need to be transmitted across await boundaries.

Agree it's just a different way to do things, and the other way was more valid in the context of regular async rust. This approach only works because tower exposes a mut call.

Reply:
Just to give a broader reason as to why clone service is undesirable: Various tower layers will convert a clonable service to a non clonable service. Requiring a clonable service means that a buffer layer has to immediately follow the layer that requires clone.